### PR TITLE
feature(nanocld): accept optional ssl keys

### LINKED ISSUE #389: Feature: Accept optional ssl keys for nanocld
**Is your feature request related to a problem? Please describe.**
We need to add optional arguments to use ssl keys for the daemon nanocld.

### PR DESCRIPTION
**Pull request by Abhicodes-crypto**
**- What I did**
I added the optional ssl params 

**- How I did it**
Added it in the DaemonState struct 

**- How to verify it**
I added tests

**- Description for the changelog**

Closes #389 

**- A picture of a cute animal (not mandatory but encouraged)**

image

### CODE REVIEW

**Comment by Abhicodes-crypto** [comment]
Please don't review yet. I will make some additional changes.

**Comment by Abhicodes-crypto** [comment]
It is ready for review now

File: bin/nanocld/src/cli.rs [hunk 1]
```diff
@@ -37,6 +37,15 @@ pub struct Cli {
   /// Group id
   #[clap(long, default_value = "0")]
   pub(crate) gid: u32,
+  /// Docker daemon ssl certificate
```
Comment (0xle0ne) [inline bin/nanocld/src/cli.rs]: It's not the docker daemon ssl keys it's ours. Just remove the `Docker daemon` part in the commentaries please!

**Review by 0xle0ne — CHANGES_REQUESTED** [review]

**Comment by 0xle0ne** [comment]
Hey @Abhicodes-crypto thanks for your contribution! It's a good start!
But you have to make the server inside `bin/nanocl/src/server.rs` use the keys now !
You have an example on how to do it [there](https://github.com/ntex-rs/examples/blob/master/openssl/src/main.rs)

File: bin/nanocld/src/cli.rs (hunk 1 above)
Comment (Abhicodes-crypto) [reply bin/nanocld/src/cli.rs]: I removed the `Docker daemon` part

Comment (0xle0ne) [reply bin/nanocld/src/cli.rs]: Thanks!

**Comment by Abhicodes-crypto** [comment]
> But you have to make the server inside bin/nanocl/src/server.rs use the keys now !
You have an example on how to do it [there](https://github.com/ntex-rs/examples/blob/master/openssl/src/main.rs)

Sure @leon3s .Would be happy to add the changes. Should I do this in the current PR or raise a new PR. I would personally prefer to raise a new PR. Because I am unsure of how to do it and would need some time. 

Some doubts that I have :
- Referring the code [here](https://github.com/ntex-rs/examples/blob/master/openssl/src/main.rs) I can pass `ssl_key` as `key.pem`, `ssl_ca` as chain file, but where do I use ssl_cert I have no idea. 
- Also in `bind_openssl` what is the port that I have to use. 
- Also curious to know the use case here.

**Comment by 0xle0ne** [comment]
Sorry @Abhicodes-crypto you have to include it in this pull request unless we are accepting the arguments without using them.

From this [code](https://github.com/ntex-rs/examples/blob/master/openssl/src/main.rs) you can see there is 2 method used:
- `set_private_key_file("key.pem")` refer to our `ssl_key`
- `set_certificate_chain_file("cert.pem")` refer to our `ssl_cert`

For the `CA` you need to use another method `set_ca_file`.

You will bind these certificate for each `tcp` host passed.

```rust
    } else if host.starts_with("tcp://") {
      let addr = host.replace("tcp://", "");
      // load ssl keys
     // if the ssl_keys are set
      let mut builder =
        SslAcceptor::mozilla_intermediate(SslMethod::tls()).unwrap();
      builder
        .set_private_key_file("tests/server.pem", SslFiletype::PEM)
        .unwrap();
      builder
        .set_certificate_chain_file("tests/cert.pem")
        .unwrap();
      builder.set_ca_file("tests/ca.pem").unwrap();
      server = match server.bind_openssl(&addr, builder) {
        Err(err) => {
          log::error!("Error binding to tcp host {}: {}", &addr, &err);
          return Err(err);
        }
        Ok(server) => server,
      };
      /// Unless we do normal binding
      server = match server.bind(&addr) {
        Err(err) => {
          log::error!("Error binding to tcp host {}: {}", &addr, &err);
          return Err(err);
        }
        Ok(server) => server,
      };
      log::info!("Listening on {}", &host);
    }
```

For the use case, it's to enable encryption between a client and a server. Where the client will need to have a specific certificate signed by the `CA` to be able to connect. It's not like classic https there the objectif it's to enable authorization aswell!
This is required for the daemon to talk to each other over internet in a secure way!

**Pull request closed without merging**